infra/discovery/src/listener/listener.js

```diff
@@ -101,66 +106,90 @@ const config = {
-      const events = [].concat(...eventArrays.filter(x => x))
-      // Filter to only new events
-      const newEvents = events.filter(
-        event => event.blockNumber > processedToBlock
-      )
-      logger.debug(`Got ${newEvents.length} new events`)
-
+  do {
+    start = new Date()
+
+    // Compute the range of blocks to process,
+    // while respecting trailing block configuration.
+    const currentBlock = await context.web3.eth.getBlockNumber()
+    let toBlock = Math.max(currentBlock - context.config.trailBlocks, 0)
+    toBlock = Math.min(toBlock, processedToBlock + 10000) // At most process a chunk of 10k blocks at once.
+
+    // Listener is up to date. Nothing to do.
+    if (toBlock <= processedToBlock) {
+      logger.debug('No new blocks to process')
+      continue
+    }
+    logger.debug(`Querying events up to ${toBlock}`)
+
+    // Update the event caches to set their max block number.
+    contractsContext.marketplace.eventCache.updateBlock(toBlock)
+    contractsContext.identityEvents.eventCache.updateBlock(toBlock)
+
+    // Retrieve all events for the relevant contracts
+    const eventArrays = await Promise.all([
+      contractsContext.marketplace.eventCache.allEvents(),
+      contractsContext.identityEvents.eventCache.allEvents()
+    ])
+
+    // Flatten array of arrays filtering out anything undefined
+    const events = [].concat(...eventArrays.filter(x => x))
+    // Filter to only new events
+    let newEvents = events.filter(event => event.blockNumber > processedToBlock)
+    logger.debug(`Got ${newEvents.length} new events`)
+
+    if (context.config.concurrency > 1) {
+      // Concurrency greater than 1 -> Backfill mode.
       const limit = pLimit(context.config.concurrency)
       const promises = []
       newEvents.forEach(newEvent => {
         promises.push(limit(() => handleEvent(newEvent, context)))
       })
       await Promise.all(promises)
-
-      // Record state of processing
-      logger.debug(`Updating last processed block to ${toBlock}`)
-      await setLastBlock(context.config, toBlock)
-      processedToBlock = toBlock
-      blockGauge.set(toBlock)
-
-      return scheduleNextCheck()
-    })
-  }
-
-  const scheduleNextCheck = async () => {
-    const elapsed = new Date() - start
-    const delay = Math.max(checkIntervalSeconds * 1000 - elapsed, 1)
-    setTimeout(check, delay)
-  }
-
-  check()
+    } else {
+      // Concurrency set to 1 -> Normal operation mode.
+      // Sort events by blockNumber and logIndex to process them in order.
+      // In normal operation (not backfill), ordering does matter for handlers.
+      // For example assume a buyer updates their identity to add their email
+      // then makes a purchase. If the handler would process the purchase
+      // event before the identity update, we may not be able to send the buyer
+      // a confirmation email about their offer.
+      newEvents = sortBy(newEvents, ['blockNumber', 'logIndex'], ['asc', 'asc'])
```

Comment:
I think we should move this up so it happens in backfill mode as well. It's probably not a big performance hit and I think it is good if we keep the logic as close as possible for backfill vs non backfill. For example, if the listener fails and we need to run with concurrency to catch up and we are missing this logic the issue we describe in the comments will still arise.

Reply:
The problem is that there is no way to guarantee processing when using Promises, right ?
Like when Promise.all(...) is called, the events are processed in parallel without guaranteed ordering. Or maybe I'm missing some fancy JS trick ?

Reply:
Yea you are right! Maybe just something to keep in mind then